File: shared/src/enums.ts
/**
 * Closed vocabularies used across the data model. Each is declared once as a
 * `const` tuple so it can drive both a union type and runtime validation.
 */

export const MEMBER_GRADES = ['Open', 'Intermediate', 'Junior', 'Unknown'] as const;
export type MemberGrade = (typeof MEMBER_GRADES)[number];

export const MEMBER_ROLES = ['member', 'admin'] as const;
export type MemberRole = (typeof MEMBER_ROLES)[number];

/** Day a weekday programme runs on. Tuesday is the juniors' evening. */
export const WEEKDAYS = ['monday', 'tuesday', 'wednesday', 'thursday', 'friday'] as const;
export type Weekday = (typeof WEEKDAYS)[number];

export const SCORING_TYPES = ['Scr', 'Hcp'] as const;
export type ScoringType = (typeof SCORING_TYPES)[number];

export const SERIES_FORMATS = ['Pairs', 'Teams', 'Individual'] as const;
export type SeriesFormat = (typeof SERIES_FORMATS)[number];

/** What a single dated session is. */
export const SESSION_KINDS = ['series', 'holidayBridge', 'noBridge'] as const;
export type SessionKind = (typeof SESSION_KINDS)[number];

/** Whether a partner reference points at a member or a non-member visitor. */
export const PARTNER_KINDS = ['member', 'visitor'] as const;
export type PartnerKind = (typeof PARTNER_KINDS)[number];

/**
 * State of one member's entry for one session.
 * - `confirmed`            paired (member or visitor), or a member of a team
 * - `looking_for_partner`  public, first-claim-wins
 * - `available`            public, a claim sends an invite
 * - `unavailable`          solo, like `available`/`looking_for_partner` (I6) but
 *                          never shown on the noticeboard and never alerted on
 *                          (plan §21 B2) — "don't offer me this session". It
 *                          still occupies the member's slot: every "is this
 *                          member free" precondition (`isFree`, §7) treats it
 *                          exactly like `confirmed`/`substituted` — only
 *                          `cancelled`/absent frees a slot.
 * - `substituted`          paired, but covered this session by a substitute
 * - `cancelled`            withdrawn; kept for history, treated as absent
 */
export const ENTRY_STATUSES = [
  'confirmed',
  'looking_for_partner',
  'available',
  'unavailable',
  'substituted',
  'cancelled',
] as const;
export type EntryStatus = (typeof ENTRY_STATUSES)[number];

/** The three solo (unpaired, I6) statuses — `partner`/`pairingId`/substitution fields are always null. */
export const SOLO_ENTRY_STATUSES = [
  'looking_for_partner',
  'available',
  'unavailable',
] as const satisfies readonly EntryStatus[];

/**
 * Statuses that occupy a place on a member's *card display* (web `My Card`,
 * `lib/card.ts#isActiveCardEntry`) — deliberately **excludes** `unavailable`,
 * which is a "don't ask me" marker, not a booking, and must not show as a row
 * there (plan §21 B2). This is a display concern only: server-side "is this
 * member free for session S" preconditions never consult this constant —
 * they use `entries/lib.ts#isFree` (any non-cancelled entry, including
 * `unavailable`, blocks), which is intentionally the wider set.
 */
export const ACTIVE_ENTRY_STATUSES = [
  'confirmed',
  'looking_for_partner',
  'available',
  'substituted',
] as const satisfies readonly EntryStatus[];

/** Statuses visible on the public noticeboard. Deliberately excludes `unavailable`. */
export const NOTICEBOARD_STATUSES = [
  'looking_for_partner',
  'available',
] as const satisfies readonly EntryStatus[];

export const INVITE_STATUSES = [
  'pending',
  'accepted',
  'declined',
  'cancelled',
  'expired',
] as const;
export type InviteStatus = (typeof INVITE_STATUSES)[number];

/** What an invite covers: a single session, a whole series sign-up, or a team invite. */
export const INVITE_SCOPES = ['session', 'series', 'team'] as const;
export type InviteScope = (typeof INVITE_SCOPES)[number];

export const PROGRAMME_STATUSES = ['draft', 'published'] as const;
export type ProgrammeStatus = (typeof PROGRAMME_STATUSES)[number];

export const TEAM_STATUSES = ['forming', 'active', 'disbanded'] as const;
export type TeamStatus = (typeof TEAM_STATUSES)[number];

/**
 * What a `scope:'team'` invite is for: joining the roster (`'join'`, the
 * default when the field is absent — every invite created before this kind
 * existed is implicitly a join invite) or accepting the captaincy
 * (`'captaincy'`, plan §9.2 `transferCaptaincy`).
 */
export const INVITE_KINDS = ['join', 'captaincy'] as const;
export type InviteKind = (typeof INVITE_KINDS)[number];

export const NOTIFICATION_CHANNELS = ['inapp', 'push', 'email', 'sms'] as const;
export type NotificationChannel = (typeof NOTIFICATION_CHANNELS)[number];

export const NOTIFICATION_TYPES = [
  'invite_received',
  'invite_accepted',
  'invite_declined',
  'invite_cancelled',
  'invite_expired',
  'claimed',
  'partner_cancelled',
  'substitute_arranged',
  'substitute_cleared',
  'matchmaking_alert',
  'session_reminder',
  'on_behalf_action',
  'team_invite_received',
  'team_member_joined',
  'team_member_declined',
  'team_member_left',
  'team_member_absent',
  'team_removed',
  'team_captaincy_offered',
  'team_captaincy_transferred',
  'team_disbanded',
  'broadcast',
  'security',
  /** §12.5: a visitor's email now matches a newly-imported member. */
  'visitor_promoted',
  /** Phase 6: `updateSeries`/`updateSession` changed something affecting a member's future entry. */
  'programme_changed',
] as const;
export type NotificationType = (typeof NOTIFICATION_TYPES)[number];

export const DIGEST_MODES = ['immediate', 'daily'] as const;
export type DigestMode = (typeof DIGEST_MODES)[number];

/**
 * Actions recorded in the audit log: admin-initiated system actions, plus one
 * `<callable>_on_behalf` action per callable that supports `onBehalfOfMemberId`
 * (§9.2), logged whenever it is actually used on someone else's behalf.
 */
export const AUDIT_ACTIONS = [
  // System / admin actions
  'member_import',
  'role_changed',
  'member_deactivated',
  'member_reactivated',
  'member_erased',
  'programme_import',
  'programme_publish',
  'programme_edit',
  'broadcast_sent',
  'pairing_repair',
  /** System action: `importMembers` promoted a visitor whose email matched a new member (§12.5). */
  'visitor_promoted',
  /** System action: `purgeExpired` deleted a visitor unused for 18+ months with no future entries. */
  'visitor_purged',
  // Per-callable on-behalf actions (§9.2 "Audit" column)
  'send_invite_on_behalf',
  'respond_to_invite_on_behalf',
  'cancel_invite_on_behalf',
  'set_solo_status_on_behalf',
  'set_bulk_solo_status_on_behalf',
  'claim_looking_for_partner_on_behalf',
  'create_visitor_on_behalf',
  'sign_up_with_visitor_on_behalf',
  'set_substitute_on_behalf',
  'clear_substitute_on_behalf',
  'cancel_entry_on_behalf',
  'create_team_on_behalf',
  'invite_to_team_on_behalf',
  'add_visitor_to_team_on_behalf',
  'remove_visitor_from_team_on_behalf',
  'leave_team_on_behalf',
  'remove_from_team_on_behalf',
  'transfer_captaincy_on_behalf',
  'disband_team_on_behalf',
  'add_team_session_substitute_on_behalf',
  'clear_team_session_substitute_on_behalf',
] as const;
export type AuditAction = (typeof AUDIT_ACTIONS)[number];

/** Runtime membership test for any of the const tuples above. */
export function isOneOf<T extends readonly string[]>(
  vocab: T,
  value: unknown,
): value is T[number] {
  return typeof value === 'string' && (vocab as readonly string[]).includes(value);
}

File: ios/OBCDanceCard/Shared/Enums.swift
//
//  Enums.swift
//  Closed vocabularies, mirrored by hand from `shared/src/enums.ts` — the
//  single source of truth (plan §3.2). Raw values are byte-for-byte identical
//  to the TypeScript const tuples; changing one here without changing the
//  other silently breaks decoding of live documents.
//
//  Every enum is `String`-backed and decodes leniently where the server may
//  add a value later (see `NotificationType`), so an unknown value from a
//  newer backend degrades to "unknown" rather than failing the whole
//  snapshot.
//

import Foundation

/// `MEMBER_GRADES`
/// App-Store-review cohort partition (plan §8.1, decided 2026-09-05):
/// `club` is every real member; `review` only the fake accounts provisioned
/// for Apple's reviewers. Rules require every roster-style read to be scoped
/// to the caller's own cohort, so the directory, teams and session-entries
/// listeners all filter on it. Mirrors `MEMBER_COHORTS` in `shared/src/enums.ts`.
enum MemberCohort: String, Codable, Hashable {
    case club, review
}

enum MemberGrade: String, Codable, CaseIterable, Hashable {
    case open = "Open"
    case intermediate = "Intermediate"
    case junior = "Junior"
    case unknown = "Unknown"
}

/// `MEMBER_ROLES`
enum MemberRole: String, Codable, Hashable {
    case member
    case admin
}

/// `WEEKDAYS` — the club runs Monday–Friday only.
enum Weekday: String, Codable, CaseIterable, Hashable {
    case monday, tuesday, wednesday, thursday, friday

    /// Short tab label, mirroring `web/src/lib/format.ts#shortWeekdayLabel`.
    var shortLabel: String {
        switch self {
        case .monday: return "Mon"
        case .tuesday: return "Tue"
        case .wednesday: return "Wed"
        case .thursday: return "Thu"
        case .friday: return "Fri"
        }
    }
}

/// `SCORING_TYPES`
enum ScoringType: String, Codable, Hashable {
    case scratch = "Scr"
    case handicap = "Hcp"
}

/// `SERIES_FORMATS`
enum SeriesFormat: String, Codable, Hashable {
    case pairs = "Pairs"
    case teams = "Teams"
    case individual = "Individual"
}

/// `SESSION_KINDS`
enum SessionKind: String, Codable, Hashable {
    case series
    case holidayBridge
    case noBridge
}

/// `PARTNER_KINDS`
enum PartnerKind: String, Codable, Hashable {
    case member
    case visitor
}

/// `ENTRY_STATUSES` — see `shared/src/enums.ts` for what each one means.
enum EntryStatus: String, Codable, Hashable {
    case confirmed
    case lookingForPartner = "looking_for_partner"
    case available
    /// Plan §21 B2: a solo "don't offer me this session" marker. Never on
    /// the noticeboard, never alerted on, never a row on the card — but it
    /// still occupies the member's slot server-side like a booking does.
    case unavailable
    case substituted
    case cancelled

    /// `ACTIVE_ENTRY_STATUSES`: statuses that occupy a place on the member's
    /// *card display*. Deliberately excludes `unavailable` (a marker, not a
    /// booking). Server-side "is this member free" checks use the wider
    /// "any non-cancelled entry" rule instead.
    static let active: Set<EntryStatus> = [.confirmed, .lookingForPartner, .available, .substituted]

    /// `SOLO_ENTRY_STATUSES`: the three unpaired statuses (I6).
    static let solo: Set<EntryStatus> = [.lookingForPartner, .available, .unavailable]

    /// `NOTICEBOARD_STATUSES`: statuses visible on the public noticeboard.
    static let noticeboard: [EntryStatus] = [.lookingForPartner, .available]
}

/// The statuses a member can set on themselves (`setSoloStatus`).
enum SoloStatus: String, Codable, Hashable, CaseIterable {
    case lookingForPartner = "looking_for_partner"
    case available
    case unavailable

    var entryStatus: EntryStatus {
        switch self {
        case .lookingForPartner: return .lookingForPartner
        case .available: return .available
        case .unavailable: return .unavailable
        }
    }
}

/// `setBulkSoloStatus`'s status argument (plan §21 B2): the two markers, or
/// `clear` to remove either (entries flip to `cancelled`; never deleted).
enum BulkAvailabilityStatus: String, Codable, Hashable, CaseIterable {
    case available, unavailable, clear
}

/// `INVITE_STATUSES`
enum InviteStatus: String, Codable, Hashable {
    case pending, accepted, declined, cancelled, expired
}

/// `INVITE_SCOPES`
enum InviteScope: String, Codable, Hashable {
    case session, series, team
}

/// `INVITE_KINDS` — absent means `.join` (plan §5.7).
enum InviteKind: String, Codable, Hashable {
    case join, captaincy
}

/// `PROGRAMME_STATUSES`
enum ProgrammeStatus: String, Codable, Hashable {
    case draft, published
}

/// `TEAM_STATUSES`
enum TeamStatus: String, Codable, Hashable {
    case forming, active, disbanded
}

/// `NOTIFICATION_CHANNELS`
enum NotificationChannel: String, Codable, Hashable {
    case inapp, push, email, sms
}

/// `DIGEST_MODES`
enum DigestMode: String, Codable, Hashable {
    case immediate, daily
}

/// `RegisteredDevice.platform`
enum DevicePlatform: String, Codable, Hashable {
    case ios, web
}

/// Which side of a pairing a substitute stands in for (`SetSubstituteInput.coverFor`).
enum CoverFor: String, Codable, Hashable {
    case selfMember = "self"
    case partner
}

/// `NOTIFICATION_TYPES`. Decoded leniently: the server owns this list and may
/// grow it, and a notification whose type this build doesn't recognise must
/// still render its title/body rather than break the whole feed.
enum NotificationType: String, Codable, Hashable {
    case inviteReceived = "invite_received"
    case inviteAccepted = "invite_accepted"
    case inviteDeclined = "invite_declined"
    case inviteCancelled = "invite_cancelled"
    case inviteExpired = "invite_expired"
    case claimed
    case partnerCancelled = "partner_cancelled"
    case substituteArranged = "substitute_arranged"
    case substituteCleared = "substitute_cleared"
    case matchmakingAlert = "matchmaking_alert"
    case sessionReminder = "session_reminder"
    case onBehalfAction = "on_behalf_action"
    case teamInviteReceived = "team_invite_received"
    case teamMemberJoined = "team_member_joined"
    case teamMemberDeclined = "team_member_declined"
    case teamMemberLeft = "team_member_left"
    case teamMemberAbsent = "team_member_absent"
    case teamRemoved = "team_removed"
    case teamCaptaincyOffered = "team_captaincy_offered"
    case teamCaptaincyTransferred = "team_captaincy_transferred"
    case teamDisbanded = "team_disbanded"
    case broadcast
    case security
    case visitorPromoted = "visitor_promoted"
    case programmeChanged = "programme_changed"
    /// Not a server value — the fallback for a type this build predates.
    case unrecognised = "__unrecognised__"

    init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = NotificationType(rawValue: raw) ?? .unrecognised
    }
}
